internal/pubsub/example_test.go

```diff
@@ -101,3 +102,141 @@ func Example_sendReceiveMultipleMessages() {
 	// b
 	// c
 }
+
+func Example_receiveWithInvertedWorkerPool() {
+	// Open a topic and corresponding subscription.
+	ctx := context.Background()
+	b := mempubsub.NewBroker([]string{"myTopic"})
+	t := mempubsub.OpenTopic(b, "myTopic")
+	defer t.Close()
+	s := mempubsub.OpenSubscription(b, "myTopic", time.Second)
+	defer s.Close()
+
+	// Send a bunch of messages to the topic.
+	const nMessages = 100
+	var wg sync.WaitGroup
+	wg.Add(nMessages)
+	for n := 0; n < nMessages; n++ {
+		m := &pubsub.Message{
+			Body: []byte(fmt.Sprintf("message %d", n)),
+		}
+		if err := t.Send(ctx, m); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	// In order to make our test exit, we keep track of how many messages were
+	// processed and cancel the receiveCtx when we've processed them all.
+	receiveCtx, cancel := context.WithCancel(ctx)
+	go func() {
+		wg.Wait()
+		cancel()
+	}()
+
+	// Process messages using an inverted worker pool, as described here:
+	// https://www.youtube.com/watch?v=5zXAHh5tJqQ&t=26m58s
+	// It uses a buffered channel, sem, as a semaphore to manage the maximum
+	// number of workers.
+	const poolSize = 10
+	sem := make(chan struct{}, poolSize)
+	for {
+		// Read a message. Receive will block until a message is available.
+		msg, err := s.Receive(receiveCtx)
+		if err != nil {
+			// An error from Receive is fatal; Receive will never succeed again
+			// so the application should exit.
+			// In this example, we expect to get a error here when we've read all the
+			// messages and receiveCtx is canceled.
+			break
+		}
+
+		// Write a token to the semaphore; if there are already poolSize workers
+		// active, this will block until one of them completes.
+		sem <- struct{}{}
+		// Process the message. For many applications, this can be expensive, so
+		// we do it in a goroutine, allowing this loop to continue and Receive more
+		// messages.
+		go func() {
+			// Record that we've processed this message, and Ack it.
+			msg.Ack()
+			wg.Done()
+			// Read a token from the semaphore before exiting this goroutine, freeing
+			// up the slot for another goroutine.
+			<-sem
+		}()
+	}
+
+	// Wait for all workers to finish.
```

Comment:
I don't believe this code is waiting for workers to finish. I think it's allowing them to start.

Reply:
I don't think so. If no workers ran at all, then this would be able to write `poolSize` tokens to `sem` immediately. This is supposed to block while any workers are running.